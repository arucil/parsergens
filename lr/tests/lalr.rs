use insta::assert_snapshot;

mod parse;

#[test]
fn precedence() {
  let parser = lr::build(r#"
%token plus "+"
%token minus "-"
%token mult "*"
%token div "/"
%token pow "^"
%token equal "=="
%token lparen "("
%token rparen ")"
%token comma ","
%token num /\d+/
%token id /[a-zA-Z]\w*/

%skip /[ \t\n]+/

%non-assoc equal
%left-assoc plus minus
%left-assoc mult div
%right-assoc pow
%right-assoc NEG

%start E

E = E plus E
  | E minus E
  | E mult E
  | E div E
  | E pow E
  | E equal E
  | minus E     %prec NEG
  | lparen E rparen
  | id lparen ARGS rparen
  | num
  | id

ARGS = ()
  | E (comma E)*
  "#, lr::ParserKind::Lalr).unwrap();

  let input = "12 + 3 * x / 5 - 50 - foo ^ -y ^ (-1 == f() + g(3*-4, (p)))";

  assert_snapshot!(parse::parse(&parser, input, "E").join("\n"));
}
