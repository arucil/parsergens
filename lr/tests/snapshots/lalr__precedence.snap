---
source: lr/tests/lalr.rs
expression: "parse::parse(&parser, input, \"E\").join(\"\\n\")"
---
shift  12
reduce E -> 12
shift  +
shift  3
reduce E -> 3
shift  *
shift  x
reduce E -> x
reduce E -> E * E
shift  /
shift  5
reduce E -> 5
reduce E -> E / E
reduce E -> E + E
shift  -
shift  50
reduce E -> 50
reduce E -> E - E
shift  -
shift  foo
reduce E -> foo
shift  ^
shift  -
shift  y
reduce E -> y
reduce E -> - E
shift  ^
shift  (
shift  -
shift  1
reduce E -> 1
reduce E -> - E
shift  ==
shift  f
shift  (
reduce ARGS ->
shift  )
reduce E -> f ( ARGS )
shift  +
shift  g
shift  (
shift  3
reduce E -> 3
shift  *
shift  -
shift  4
reduce E -> 4
reduce E -> - E
reduce E -> E * E
reduce (comma E)* ->
shift  ,
shift  (
shift  p
reduce E -> p
shift  )
reduce E -> ( E )
reduce (comma E)* -> (comma E)* , E
reduce ARGS -> E (comma E)*
shift  )
reduce E -> g ( ARGS )
reduce E -> E + E
reduce E -> E == E
shift  )
reduce E -> ( E )
reduce E -> E ^ E
reduce E -> E ^ E
reduce E -> E - E
accept
